crates/accelerate/src/commutation_checker.rs

```diff
@@ -0,0 +1,738 @@
+                first_mat
+            };
+            let op12 = compose(&first_mat.view(), &second_mat.view(), &second_qarg, false);
+            let op21 = compose(&first_mat.view(), &second_mat.view(), &second_qarg, true);
+            abs_diff_eq!(op12, op21, epsilon = 1e-8)
+        }
+    }
+
+    fn clear_cache(&mut self) {
+        self.cache.clear();
+        self.current_cache_entries = 0;
+        self._cache_miss = 0;
+        self._cache_hit = 0;
+    }
+}
+
+#[allow(clippy::too_many_arguments)]
+fn commutation_precheck(
+    op1: &CircuitInstruction,
+    qargs1: &[usize],
+    cargs1: &[usize],
+    op2: &CircuitInstruction,
+    qargs2: &[usize],
+    cargs2: &[usize],
+    max_num_qubits: u32,
+) -> Option<bool> {
+    if op1.op().control_flow()
+        || op2.op().control_flow()
+        || op1.is_conditioned()
+        || op2.is_conditioned()
+    {
+        return Some(false);
+    }
+
+    // assuming the number of involved qubits to be small, this might be faster than set operations
+    if !qargs1.iter().any(|e| qargs2.contains(e)) && !cargs1.iter().any(|e| cargs2.contains(e)) {
+        return Some(true);
+    }
+
+    if qargs1.len() > max_num_qubits as usize || qargs2.len() > max_num_qubits as usize {
+        return Some(false);
+    }
+
+    if SUPPORTED_OP.contains(&op1.op().name()) && SUPPORTED_OP.contains(&op2.op().name()) {
+        return None;
+    }
+
+    if is_commutation_skipped(op1) || is_commutation_skipped(op2) {
+        return Some(false);
+    }
+
+    None
+}
+
+fn get_op(gate: &PyObject) -> Option<Array2<Complex64>> {
+    Python::with_gil(|py| -> Option<Array2<Complex64>> {
+        Some(
+            QI_OPERATOR
+                .get_bound(py)
+                .call1((gate,))
+                .ok()?
+                .getattr(intern!(py, "data"))
+                .ok()?
+                .extract::<PyReadonlyArray2<Complex64>>()
+                .ok()?
+                .as_array()
+                .to_owned(),
+        )
+    })
+}
+
+fn is_commutation_skipped(instr: &CircuitInstruction) -> bool {
+    let op = instr.op();
+    op.directive() || SKIPPED_NAMES.contains(&op.name()) || instr.is_parameterized()
+}
+
+fn get_relative_placement(
+    first_qargs: &[usize],
+    second_qargs: &[usize],
+) -> SmallVec<[Option<Qubit>; 2]> {
+    let qubits_g2: HashMap<_, _> = second_qargs
+        .iter()
+        .enumerate()
+        .map(|(i_g1, q_g1)| (q_g1, Qubit(i_g1 as u32)))
+        .collect();
+
+    first_qargs
+        .iter()
+        .map(|q_g0| qubits_g2.get(q_g0).copied())
+        .collect()
+}
+
+#[derive(Clone, Debug)]
+#[pyclass]
+pub struct CommutationLibrary {
+    pub library: HashMap<(String, String), CommutationLibraryEntry>,
+}
+
+impl CommutationLibrary {
+    fn check_commutation_entries(
+        &self,
+        first_op: &OperationRef,
+        first_qargs: &[usize],
+        second_op: &OperationRef,
+        second_qargs: &[usize],
+    ) -> Option<bool> {
```

Comment:
Use `u32` here instead.

```suggestion
    fn check_commutation_entries(
        &self,
        first_op: &OperationRef,
        first_qargs: &[u32],
        second_op: &OperationRef,
        second_qargs: &[u32],
    ) -> Option<bool> {
```